File: sdks/typescript/src/worker.ts
import { randomUUID } from "node:crypto";
import { createChannel, createClient, type Channel } from "nice-grpc";
import { TaskContext } from "./context.js";
import { ConnectionError, HandlerError } from "./errors.js";
import {
  WorkerServiceDefinition,
  type DeepPartial,
  type WorkerRequest,
} from "./generated/valka/v1/worker.js";
import { RetryPolicy } from "./retry.js";
import type { TaskHandler, ValkaWorkerOptions } from "./types.js";

interface MessageChannel<T> {
  send(msg: T): void;
  close(): void;
  [Symbol.asyncIterator](): AsyncIterableIterator<T>;
}

function createMessageChannel<T>(): MessageChannel<T> {
  const queue: T[] = [];
  let resolve: (() => void) | null = null;
  let done = false;

  return {
    send(msg: T) {
      queue.push(msg);
      if (resolve) {
        resolve();
        resolve = null;
      }
    },
    close() {
      done = true;
      if (resolve) {
        resolve();
        resolve = null;
      }
    },
    [Symbol.asyncIterator](): AsyncIterableIterator<T> {
      const iter: AsyncIterableIterator<T> = {
        async next(): Promise<IteratorResult<T>> {
          while (queue.length === 0 && !done) {
            await new Promise<void>((r) => {
              resolve = r;
            });
          }
          if (queue.length > 0) {
            return { value: queue.shift()!, done: false };
          }
          return { value: undefined as unknown as T, done: true };
        },
        [Symbol.asyncIterator]() {
          return iter;
        },
      };
      return iter;
    },
  };
}

export class ValkaWorker {
  private readonly workerId: string;
  private readonly name: string;
  private readonly serverAddr: string;
  private readonly queues: string[];
  private readonly concurrency: number;
  private readonly handler: TaskHandler;
  private readonly metadataStr: string;
  private shutdownResolve: (() => void) | null = null;
  private readonly shutdownPromise: Promise<void>;

  private constructor(options: ValkaWorkerOptions) {
    this.workerId = randomUUID();
    this.name = options.name ?? `worker-${randomUUID().slice(0, 8)}`;
    this.serverAddr = options.serverAddr ?? "localhost:50051";
    this.queues = options.queues;
    this.concurrency = options.concurrency ?? 1;
    this.handler = options.handler;
    this.metadataStr = options.metadata ? JSON.stringify(options.metadata) : "";
    this.shutdownPromise = new Promise((resolve) => {
      this.shutdownResolve = resolve;
    });
  }

  shutdown(): void {
    if (this.shutdownResolve) {
      this.shutdownResolve();
      this.shutdownResolve = null;
    }
  }

  static builder(): ValkaWorkerBuilder {
    return new ValkaWorkerBuilder();
  }

  static create(options: ValkaWorkerOptions): ValkaWorker {
    return new ValkaWorker(options);
  }

  async run(): Promise<void> {
    const retryPolicy = new RetryPolicy();

    // eslint-disable-next-line no-constant-condition
    while (true) {
      try {
        await this.connectAndRun(retryPolicy);
        console.log("[valka] Worker disconnected gracefully");
        return;
      } catch (err) {
        const delay = retryPolicy.nextDelay();
        console.warn(
          `[valka] Connection lost: ${err instanceof Error ? err.message : err}. ` +
            `Reconnecting in ${delay}ms...`,
        );
        await sleep(delay);
      }
    }
  }

  private async connectAndRun(retryPolicy: RetryPolicy): Promise<void> {
    const channel: Channel = createChannel(this.serverAddr);
    const client = createClient(WorkerServiceDefinition, channel);
    const requests = createMessageChannel<DeepPartial<WorkerRequest>>();

    retryPolicy.reset();
    console.log(
      `[valka] Connected: worker_id=${this.workerId} name=${this.name} ` +
        `queues=[${this.queues.join(",")}] concurrency=${this.concurrency}`,
    );

    let gracefulShutdown = false;
    const activeTasks = new Set<string>();
    const taskContexts = new Map<string, TaskContext>();
    let activeCount = 0;
    const pendingResolves: Array<() => void> = [];

    // Semaphore-like concurrency control
    const acquireSlot = (): Promise<void> => {
      if (activeCount < this.concurrency) {
        activeCount++;
        return Promise.resolve();
      }
      return new Promise<void>((res) => pendingResolves.push(res));
    };
    const releaseSlot = (): void => {
      if (pendingResolves.length > 0) {
        const next = pendingResolves.shift()!;
        next();
      } else {
        activeCount--;
      }
    };

    const send = (msg: DeepPartial<WorkerRequest>): void => {
      requests.send(msg);
    };

    // Send WorkerHello
    send({
      hello: {
        workerId: this.workerId,
        workerName: this.name,
        queues: this.queues,
        concurrency: this.concurrency,
        metadata: this.metadataStr,
      },
    });

    // Heartbeat every 10 seconds
    const heartbeatInterval = setInterval(() => {
      send({
        heartbeat: {
          activeTaskIds: Array.from(activeTasks),
          timestampMs: Date.now(),
        },
      });
    }, 10_000);

    // Graceful shutdown on signals
    const shutdownHandler = () => {
      if (gracefulShutdown) return;
      console.log("[valka] Shutdown signal received, draining...");
      gracefulShutdown = true;
      send({ shutdown: { reason: "SIGINT" } });

      const checkDrained = setInterval(() => {
        if (activeTasks.size === 0) {
          clearInterval(checkDrained);
          requests.close();
        }
      }, 100);

      // Safety timeout: force close after 30s
      setTimeout(() => {
        clearInterval(checkDrained);
        requests.close();
      }, 30_000);
    };

    process.once("SIGINT", shutdownHandler);
    process.once("SIGTERM", shutdownHandler);
    this.shutdownPromise.then(shutdownHandler);

    const cleanup = () => {
      clearInterval(heartbeatInterval);
      process.removeListener("SIGINT", shutdownHandler);
      process.removeListener("SIGTERM", shutdownHandler);
    };

    try {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any -- ts-proto's Exact
      // constraint + TS 5.7's IteratorResult change cause a deep structural mismatch
      // with AsyncIterable generics. The runtime types are correct.
      for await (const response of client.session(requests as any)) {
        if (response.taskAssignment) {
          const assignment = response.taskAssignment;
          activeTasks.add(assignment.taskId);

          acquireSlot().then(async () => {
            const ctx = new TaskContext(
              assignment.taskId,
              assignment.taskRunId,
              assignment.queueName,
              assignment.taskName,
              assignment.attemptNumber,
              assignment.input,
              assignment.metadata,
              send,
            );
            taskContexts.set(assignment.taskId, ctx);

            try {
              const output = await this.handler(ctx);
              send({
                taskResult: {
                  taskId: assignment.taskId,
                  taskRunId: assignment.taskRunId,
                  success: true,
                  retryable: false,
                  output: output != null ? JSON.stringify(output) : "",
                  errorMessage: "",
                },
              });
            } catch (err) {
              const isRetryable = err instanceof HandlerError ? err.retryable : true;
              const errorMessage = err instanceof Error ? err.message : String(err);
              send({
                taskResult: {
                  taskId: assignment.taskId,
                  taskRunId: assignment.taskRunId,
                  success: false,
                  retryable: isRetryable,
                  output: "",
                  errorMessage,
                },
              });
            } finally {
              activeTasks.delete(assignment.taskId);
              taskContexts.delete(assignment.taskId);
              releaseSlot();
            }
          });
        } else if (response.taskSignal) {
          const signal = response.taskSignal;
          const ctx = taskContexts.get(signal.taskId);
          if (ctx) {
            ctx._deliverSignal(signal);
          }
        } else if (response.taskCancellation) {
          const cancel = response.taskCancellation;
          console.log(`[valka] Task cancelled: ${cancel.taskId} (${cancel.reason})`);
          activeTasks.delete(cancel.taskId);
          taskContexts.delete(cancel.taskId);
        } else if (response.heartbeatAck) {
          // No-op
        } else if (response.serverShutdown) {
          const shutdown = response.serverShutdown;
          console.log(`[valka] Server shutting down: ${shutdown.reason}`);
          break;
        }
      }
    } catch (err) {
      cleanup();
      if (gracefulShutdown) return;
      throw new ConnectionError(err instanceof Error ? err.message : String(err));
    }

    cleanup();
    requests.close();
    channel.close();
  }
}

// --- Builder ---

export class ValkaWorkerBuilder {
  private options: Partial<ValkaWorkerOptions> = {};

  name(name: string): this {
    this.options.name = name;
    return this;
  }

  serverAddr(addr: string): this {
    this.options.serverAddr = addr;
    return this;
  }

  queues(queues: string[]): this {
    this.options.queues = queues;
    return this;
  }

  concurrency(n: number): this {
    this.options.concurrency = n;
    return this;
  }

  metadata(metadata: Record<string, unknown>): this {
    this.options.metadata = metadata;
    return this;
  }

  handler(fn: TaskHandler): this {
    this.options.handler = fn;
    return this;
  }

  build(): ValkaWorker {
    if (!this.options.queues || this.options.queues.length === 0) {
      throw new Error("At least one queue is required");
    }
    if (!this.options.handler) {
      throw new Error("A handler function is required");
    }
    return ValkaWorker.create(this.options as ValkaWorkerOptions);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: sdks/go/worker.go
package valka

import (
	"context"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	pb "github.com/valka-queue/valka/sdks/go/proto/valkav1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/keepalive"
)

// TaskHandler is the function signature for processing tasks.
type TaskHandler func(ctx *TaskContext) (interface{}, error)

// ValkaWorker connects to the Valka server via gRPC bidirectional streaming.
type ValkaWorker struct {
	workerID    string
	name        string
	serverAddr  string
	queues      []string
	concurrency int
	metadata    map[string]interface{}
	handler     TaskHandler

	semaphore    chan struct{}
	activeTasks  sync.Map
	wg           sync.WaitGroup
	shuttingDown bool
	shutdownMu   sync.Mutex
	shutdownCh   chan struct{}
	sendCh       chan *pb.WorkerRequest
}

// NewWorker creates a new ValkaWorker with functional options.
func NewWorker(opts ...WorkerOption) (*ValkaWorker, error) {
	cfg := defaultConfig()
	for _, opt := range opts {
		if err := opt(cfg); err != nil {
			return nil, err
		}
	}
	if err := cfg.validate(); err != nil {
		return nil, err
	}

	workerID := generateUUID()
	name := cfg.name
	if name == "" {
		name = fmt.Sprintf("go-worker-%s", workerID[:8])
	}

	return &ValkaWorker{
		workerID:    workerID,
		name:        name,
		serverAddr:  cfg.serverAddr,
		queues:      cfg.queues,
		concurrency: cfg.concurrency,
		metadata:    cfg.metadata,
		handler:     cfg.handler,
		semaphore:   make(chan struct{}, cfg.concurrency),
		shutdownCh:  make(chan struct{}),
		sendCh:      make(chan *pb.WorkerRequest, 256),
	}, nil
}

// Run starts the worker. Blocks until shutdown or unrecoverable error.
func (w *ValkaWorker) Run(ctx context.Context) error {
	// Handle OS signals
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		select {
		case <-sigCh:
			w.Shutdown()
		case <-w.shutdownCh:
		}
	}()

	retry := DefaultRetryPolicy()
	for {
		w.shutdownMu.Lock()
		done := w.shuttingDown
		w.shutdownMu.Unlock()
		if done {
			return nil
		}

		err := w.session(ctx)
		if err == nil {
			return nil
		}

		w.shutdownMu.Lock()
		done = w.shuttingDown
		w.shutdownMu.Unlock()
		if done {
			return nil
		}

		delay := retry.NextDelay()
		log.Printf("[valka] Connection lost (%v), reconnecting in %v", err, delay)
		select {
		case <-time.After(delay):
		case <-w.shutdownCh:
			return nil
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

// Shutdown initiates graceful shutdown.
func (w *ValkaWorker) Shutdown() {
	w.shutdownMu.Lock()
	if w.shuttingDown {
		w.shutdownMu.Unlock()
		return
	}
	w.shuttingDown = true
	w.shutdownMu.Unlock()

	log.Println("[valka] Shutting down, draining active tasks...")

	// Send graceful shutdown message
	w.sendCh <- &pb.WorkerRequest{
		Request: &pb.WorkerRequest_Shutdown{
			Shutdown: &pb.GracefulShutdown{
				Reason: "client shutdown",
			},
		},
	}

	// Wait up to 30s for active tasks to drain
	done := make(chan struct{})
	go func() {
		w.wg.Wait()
		close(done)
	}()

	select {
	case <-done:
		log.Println("[valka] All tasks drained")
	case <-time.After(30 * time.Second):
		log.Println("[valka] Drain timeout, forcing shutdown")
	}

	close(w.shutdownCh)
}

func (w *ValkaWorker) session(ctx context.Context) error {
	conn, err := grpc.NewClient(
		w.serverAddr,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithKeepaliveParams(keepalive.ClientParameters{
			Time:    10 * time.Second,
			Timeout: 5 * time.Second,
		}),
	)
	if err != nil {
		return NewConnectionError("failed to connect", err)
	}
	defer conn.Close()

	client := pb.NewWorkerServiceClient(conn)
	stream, err := client.Session(ctx)
	if err != nil {
		return NewConnectionError("failed to open session", err)
	}

	// Send hello
	metaStr := ""
	if w.metadata != nil {
		data, _ := json.Marshal(w.metadata)
		metaStr = string(data)
	}
	hello := &pb.WorkerRequest{
		Request: &pb.WorkerRequest_Hello{
			Hello: &pb.WorkerHello{
				WorkerId:    w.workerID,
				WorkerName:  w.name,
				Queues:      w.queues,
				Concurrency: int32(w.concurrency),
				Metadata:    metaStr,
			},
		},
	}
	if err := stream.Send(hello); err != nil {
		return NewConnectionError("failed to send hello", err)
	}

	log.Printf("[valka] Connected as %s (id=%s, queues=%v, concurrency=%d)",
		w.name, w.workerID, w.queues, w.concurrency)

	// Dedicated sender goroutine — stream.Send is not thread-safe
	senderCtx, senderCancel := context.WithCancel(ctx)
	defer senderCancel()

	go w.senderLoop(senderCtx, stream)
	go w.heartbeatLoop(senderCtx)

	// Receive loop
	for {
		resp, err := stream.Recv()
		if err != nil {
			return NewConnectionError("stream receive error", err)
		}

		w.shutdownMu.Lock()
		done := w.shuttingDown
		w.shutdownMu.Unlock()
		if done {
			return nil
		}

		w.handleResponse(ctx, resp)
	}
}

func (w *ValkaWorker) senderLoop(ctx context.Context, stream pb.WorkerService_SessionClient) {
	for {
		select {
		case <-ctx.Done():
			return
		case msg, ok := <-w.sendCh:
			if !ok {
				return
			}
			if err := stream.Send(msg); err != nil {
				log.Printf("[valka] Send error: %v", err)
				return
			}
		}
	}
}

func (w *ValkaWorker) heartbeatLoop(ctx context.Context) {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-w.shutdownCh:
			return
		case <-ticker.C:
			var activeIDs []string
			w.activeTasks.Range(func(key, _ interface{}) bool {
				activeIDs = append(activeIDs, key.(string))
				return true
			})

			w.sendCh <- &pb.WorkerRequest{
				Request: &pb.WorkerRequest_Heartbeat{
					Heartbeat: &pb.Heartbeat{
						ActiveTaskIds: activeIDs,
						TimestampMs:   time.Now().UnixMilli(),
					},
				},
			}
		}
	}
}

func (w *ValkaWorker) handleResponse(ctx context.Context, resp *pb.WorkerResponse) {
	switch msg := resp.Response.(type) {
	case *pb.WorkerResponse_TaskAssignment:
		w.handleTaskAssignment(ctx, msg.TaskAssignment)
	case *pb.WorkerResponse_TaskCancellation:
		w.handleTaskCancellation(msg.TaskCancellation)
	case *pb.WorkerResponse_ServerShutdown:
		log.Printf("[valka] Server shutdown: %s", msg.ServerShutdown.Reason)
		go w.Shutdown()
	case *pb.WorkerResponse_HeartbeatAck:
		// no-op
	}
}

func (w *ValkaWorker) handleTaskAssignment(ctx context.Context, assignment *pb.TaskAssignment) {
	// Acquire semaphore slot
	w.semaphore <- struct{}{}
	w.wg.Add(1)

	taskCtx, taskCancel := context.WithCancel(ctx)
	w.activeTasks.Store(assignment.TaskId, taskCancel)

	go func() {
		defer func() {
			<-w.semaphore
			w.activeTasks.Delete(assignment.TaskId)
			w.wg.Done()
		}()

		w.executeTask(taskCtx, taskCancel, assignment)
	}()
}

func (w *ValkaWorker) handleTaskCancellation(cancellation *pb.TaskCancellation) {
	if cancelFn, ok := w.activeTasks.Load(cancellation.TaskId); ok {
		log.Printf("[valka] Task %s cancelled: %s", cancellation.TaskId, cancellation.Reason)
		cancelFn.(context.CancelFunc)()
	}
}

func (w *ValkaWorker) executeTask(ctx context.Context, cancel context.CancelFunc, assignment *pb.TaskAssignment) {
	defer cancel()

	tctx := &TaskContext{
		Context:       ctx,
		TaskID:        assignment.TaskId,
		TaskRunID:     assignment.TaskRunId,
		QueueName:     assignment.QueueName,
		TaskName:      assignment.TaskName,
		AttemptNumber: assignment.AttemptNumber,
		RawInput:      assignment.Input,
		RawMetadata:   assignment.Metadata,
		sendFn: func(req *pb.WorkerRequest) {
			select {
			case w.sendCh <- req:
			default:
				// Drop if channel is full to avoid blocking
			}
		},
		cancel: cancel,
	}

	success := false
	retryable := true
	output := ""
	errorMessage := ""

	result, err := w.handler(tctx)
	if err != nil {
		if he, ok := err.(*HandlerError); ok {
			retryable = he.Retryable
			errorMessage = he.Error()
		} else {
			errorMessage = err.Error()
		}
		log.Printf("[valka] Task %s failed: %s", assignment.TaskId, errorMessage)
	} else {
		success = true
		if result != nil {
			data, err := json.Marshal(result)
			if err != nil {
				output = fmt.Sprintf("%v", result)
			} else {
				output = string(data)
			}
		}
	}

	w.sendCh <- &pb.WorkerRequest{
		Request: &pb.WorkerRequest_TaskResult{
			TaskResult: &pb.TaskResult{
				TaskId:       assignment.TaskId,
				TaskRunId:    assignment.TaskRunId,
				Success:      success,
				Retryable:    retryable,
				Output:       output,
				ErrorMessage: errorMessage,
			},
		},
	}
}

// generateUUID generates a UUID v4 using crypto/rand (no external deps).
func generateUUID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant 2
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
